### PR TITLE
Add secure signature generation function

### PR DESCRIPTION
**Pull request by khazamov**
Added example app for secure uploads

### CODE REVIEW

File: examples/signeduploads/sampleapp/views.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+from django.shortcuts import render
+from django.http import HttpResponse
+import json
+import time
+from .forms import ImgForm
+from pyuploadcare import generate_secure_signature
+from django.conf import settings
+from django.views.decorators.csrf import csrf_exempt
```
Comment (dmitry-mukhin) [inline examples/signeduploads/sampleapp/views.py]: pep8 on imports

File: examples/signeduploads/sampleapp/views.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+def index(request):
+    imgForm = ImgForm()
+    return render(request, 'sampleapp/index.html', {'form': imgForm})
+
+
+@csrf_exempt
+def release_signature(request):
+    # we will set the signature to expire in 30 min
+    expire = int(time.time()) + 60 * 30
+    signature = generate_secure_signature(
```
Comment (dmitry-mukhin) [inline examples/signeduploads/sampleapp/views.py]:
```
generate_secure_signature(settings.UPLOADCARE['secret'],
                          expire)
```

File: examples/signeduploads/signeduploads/urls.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+Function views
+    1. Add an import:  from my_app import views
+    2. Add a URL to urlpatterns:  url(r'^$', views.home, name='home')
+Class-based views
+    1. Add an import:  from other_app.views import Home
+    2. Add a URL to urlpatterns:  url(r'^$', Home.as_view(), name='home')
+Including another URLconf
+    1. Import the include() function: from django.conf.urls import url, include
+    2. Add a URL to urlpatterns:  url(r'^blog/', include('blog.urls'))
+"""
```
Comment (dmitry-mukhin) [inline examples/signeduploads/signeduploads/urls.py]: why do we have all this boilerplate stuff

File: examples/signeduploads/templates/sampleapp/index.html [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    padding-top: 20px;
+    padding-right: 10px;
+    padding-bottom: 10px;
+    padding-left: 10px;
+
+
+}
+
+</style>
+<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.3/jquery.min.js"></script>
```
Comment (dmitry-mukhin) [inline examples/signeduploads/templates/sampleapp/index.html]: we don't need this if we're using uploadcare.full

File: examples/signeduploads/templates/sampleapp/index.html [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+ // For simplicity we are giving signatures to all guests.
+
+ $.ajax({
+     url: 'release_signature',
+     type: 'POST',
+     success: function(response) {
+                     signature = response['signature'];
+                     expire = response['expire'];
+                     uploadcare.start({
+                         publicKey: 'dddc66243e5fad63afba',
```
Comment (dmitry-mukhin) [inline examples/signeduploads/templates/sampleapp/index.html]: ???

File: pyuploadcare/api.py [hunk 6]
```diff
@@ -17,7 +17,6 @@
 import socket
 import cgi
 import time
-
```
Comment (dmitry-mukhin) [inline pyuploadcare/api.py]: pep8 on imports

File: pyuploadcare/api.py [hunk 7]
```diff
@@ -204,7 +208,7 @@ def uploading_request(verb, path, data=None, files=None, timeout=conf.DEFAULT):
 
     Make sure that given ``path`` does not contain leading slash.
 
-    Usage example::
+    Usage example::
```
Comment (dmitry-mukhin) [inline pyuploadcare/api.py]: trailing whitespaces, please configure your editor to trim those

File: pyuploadcare/api_resources.py [hunk 8]
```diff
@@ -277,7 +277,7 @@ def construct_from(cls, file_info):
         return file_
 
     @classmethod
-    def upload(cls, file_obj, store=None):
+    def upload(cls, file_obj, store=None, signature=None, expire=None):
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: if we have `conf.signed_uploads` we don't have to change the method signature.
but we have to add an expire to conf or as an argument.

File: pyuploadcare/api.py [hunk 9]
```diff
@@ -196,6 +195,11 @@ def _request():
             else:
                 raise
 
+def generate_secure_signature(secret, expire):
```
Comment (dmitry-mukhin) [inline pyuploadcare/api.py]: 1. it shouldn't sit here. move to utils
2. both secret and expire should be optional and default to `conf.xxx`

File: examples/signeduploads/sampleapp/views.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+from django.views.decorators.csrf import csrf_exempt
+
+
+def index(request):
+    imgForm = ImgForm()
+    return render(request, 'sampleapp/index.html', {'form': imgForm})
+
+
+@csrf_exempt
+def release_signature(request):
```
Comment (dmitry-mukhin) [inline examples/signeduploads/sampleapp/views.py]: why "release"?

Comment (dmitry-mukhin) [inline examples/signeduploads/sampleapp/views.py]: this view func doesn't call for full blown project. just a `views.py` will cut it.

Comment (khazamov) [inline examples/signeduploads/sampleapp/views.py]: planning to add here auth-based views, and assign signatures to users based on who is signed-in, therefore "release_signature".

File: pyuploadcare/api.py (hunk 9 above)
Comment (khazamov) [inline pyuploadcare/api.py]: 1. agreed 2. agreed, but only secret should default to conf.xxx

File: examples/signeduploads/sampleapp/views.py (hunk 10 above)
Comment (dmitry-mukhin) [inline examples/signeduploads/sampleapp/views.py]: explain "release" semantics to me.
still we don't need a project here. users will be lost in moot code.

File: pyuploadcare/api.py (hunk 9 above)
Comment (dmitry-mukhin) [inline pyuploadcare/api.py]: 2. why

File: examples/signeduploads/sampleapp/views.py (hunk 10 above)
Comment (khazamov) [inline examples/signeduploads/sampleapp/views.py]: 1. The idea is that server releases signature by request from logged-in client.
2. Agreed. I just didn't know you could do the app separate from the project :)

File: pyuploadcare/api.py (hunk 6 above)
Comment (khazamov) [inline pyuploadcare/api.py]: done

File: examples/signeduploads/sampleapp/views.py (hunk 10 above)
Comment (khazamov) [inline examples/signeduploads/sampleapp/views.py]: fixed

File: pyuploadcare/api.py (hunk 9 above)
Comment (khazamov) [inline pyuploadcare/api.py]: fixed

File: pyuploadcare/api.py (hunk 7 above)
Comment (khazamov) [inline pyuploadcare/api.py]: fixed

File: pyuploadcare/api_resources.py (hunk 8 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: fixed

File: examples/signeduploads/templates/sampleapp/index.html (hunk 5 above)
Comment (khazamov) [inline examples/signeduploads/templates/sampleapp/index.html]: fixed

File: examples/signeduploads/templates/sampleapp/index.html (hunk 4 above)
Comment (khazamov) [inline examples/signeduploads/templates/sampleapp/index.html]: fixed

File: examples/signeduploads/signeduploads/urls.py (hunk 3 above)
Comment (khazamov) [inline examples/signeduploads/signeduploads/urls.py]: fixed

File: examples/signeduploads/sampleapp/views.py (hunk 2 above)
Comment (khazamov) [inline examples/signeduploads/sampleapp/views.py]: fixed

File: examples/signeduploads/sampleapp/views.py (hunk 1 above)
Comment (khazamov) [inline examples/signeduploads/sampleapp/views.py]: done

**Pull request closed without merging**